File: android/src/main/kotlin/dev/sogl/liquid_ai/DownloadProgressHandler.kt
package dev.sogl.liquid_ai

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

// / Handles streaming download/load progress events to Flutter via EventChannel.
class DownloadProgressHandler : EventChannel.StreamHandler {
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val lock = Any()

    override fun onListen(
        arguments: Any?,
        events: EventChannel.EventSink?,
    ) {
        synchronized(lock) {
            eventSink = events
        }
    }

    override fun onCancel(arguments: Any?) {
        synchronized(lock) {
            eventSink = null
        }
    }

    // / Sends a progress event to Flutter.
    fun sendProgress(
        operationId: String,
        type: OperationType,
        status: OperationStatus,
        progress: Double = 0.0,
        speed: Long = 0,
        runnerId: String? = null,
        error: String? = null,
        errorCode: ErrorCode? = null,
    ) {
        val sink: EventChannel.EventSink?
        synchronized(lock) {
            sink = eventSink
        }

        val event =
            mutableMapOf<String, Any>(
                "operationId" to operationId,
                "type" to type.value,
                "status" to status.value,
                "progress" to progress,
                "speed" to speed,
            )

        runnerId?.let { event["runnerId"] = it }
        error?.let { event["error"] = it }
        errorCode?.let { event["errorCode"] = it.value }

        mainHandler.post {
            sink?.success(event)
        }
    }

    // / Sends an error event.
    fun sendError(
        code: String,
        message: String,
        details: Any? = null,
    ) {
        val sink: EventChannel.EventSink?
        synchronized(lock) {
            sink = eventSink
        }

        mainHandler.post {
            sink?.error(code, message, details)
        }
    }
}

// / Type of operation being performed.
enum class OperationType(
    val value: String,
) {
    DOWNLOAD("download"),
    LOAD("load"),
}

// / Current status of the operation.
enum class OperationStatus(
    val value: String,
) {
    STARTED("started"),
    PROGRESS("progress"),
    COMPLETED("completed"),
    ERROR("error"),
    CANCELLED("cancelled"),
}

// / Error codes for categorizing errors.
enum class ErrorCode(
    val value: String,
) {
    INSUFFICIENT_MEMORY("INSUFFICIENT_MEMORY"),
    MODEL_NOT_FOUND("MODEL_NOT_FOUND"),
    NETWORK_ERROR("NETWORK_ERROR"),
    REQUEST_INTERRUPTED("REQUEST_INTERRUPTED"),
    MODEL_LOADING_FAILURE("MODEL_LOADING_FAILURE"),
    INTERNAL_ERROR("INTERNAL_ERROR"),
    NOT_IMPLEMENTED("NOT_IMPLEMENTED"),
    UNKNOWN("UNKNOWN"),
}

// / Parses an exception to determine its error code.
fun parseErrorCode(error: Throwable): ErrorCode {
    val errorString = error.toString().lowercase()
    val message = (error.message ?: "").lowercase()
    val combined = "$errorString $message"

    // Check for our custom InsufficientMemoryException first
    if (error is InsufficientMemoryException) {
        return ErrorCode.INSUFFICIENT_MEMORY
    }

    // Check for memory-related errors (OutOfMemoryError is common on Android)
    if (error is OutOfMemoryError ||
        combined.contains("memory") ||
        combined.contains("oom") ||
        combined.contains("cannot allocate") ||
        combined.contains("allocation failed") ||
        combined.contains("out of memory")
    ) {
        return ErrorCode.INSUFFICIENT_MEMORY
    }

    // Check for request interrupted
    if (combined.contains("interrupted") ||
        combined.contains("cancelled") ||
        combined.contains("canceled")
    ) {
        return ErrorCode.REQUEST_INTERRUPTED
    }

    // Check for model not found
    if (combined.contains("not found") ||
        combined.contains("does not exist") ||
        combined.contains("no such file") ||
        error is java.io.FileNotFoundException
    ) {
        return ErrorCode.MODEL_NOT_FOUND
    }

    // Check for network errors
    if (combined.contains("network") ||
        combined.contains("connection") ||
        combined.contains("timeout") ||
        combined.contains("unreachable") ||
        error is java.net.UnknownHostException ||
        error is java.net.SocketTimeoutException ||
        error is java.net.ConnectException
    ) {
        return ErrorCode.NETWORK_ERROR
    }

    // Check for internal errors
    if (combined.contains("internalerror") ||
        combined.contains("internal error")
    ) {
        return ErrorCode.INTERNAL_ERROR
    }

    // Check for model loading failures
    if (combined.contains("failed to load") ||
        combined.contains("load failed")
    ) {
        return ErrorCode.MODEL_LOADING_FAILURE
    }

    return ErrorCode.UNKNOWN
}

File: ios/liquid_ai/Sources/liquid_ai/DownloadProgressHandler.swift
import Flutter
import Foundation

/// Handles streaming download/load progress events to Flutter via EventChannel.
final class DownloadProgressHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private let lock = NSLock()

    func onListen(
        withArguments _: Any?,
        eventSink events: @escaping FlutterEventSink
    )
        -> FlutterError?
    {
        lock.lock()
        defer { lock.unlock() }
        eventSink = events
        return nil
    }

    func onCancel(withArguments _: Any?) -> FlutterError? {
        lock.lock()
        defer { lock.unlock() }
        eventSink = nil
        return nil
    }

    /// Sends a progress event to Flutter.
    func sendProgress(
        operationId: String,
        type: OperationType,
        status: OperationStatus,
        progress: Double = 0.0,
        speed: Int64 = 0,
        runnerId: String? = nil,
        error: String? = nil,
        errorCode: ErrorCode? = nil
    ) {
        lock.lock()
        let sink = eventSink
        lock.unlock()

        var event: [String: Any] = [
            "operationId": operationId,
            "type": type.rawValue,
            "status": status.rawValue,
            "progress": progress,
            "speed": speed,
        ]

        if let runnerId {
            event["runnerId"] = runnerId
        }

        if let error {
            event["error"] = error
        }

        if let errorCode {
            event["errorCode"] = errorCode.rawValue
        }

        DispatchQueue.main.async {
            sink?(event)
        }
    }

    /// Sends an error event and ends the stream.
    func sendError(code: String, message: String, details: Any? = nil) {
        lock.lock()
        let sink = eventSink
        lock.unlock()

        DispatchQueue.main.async {
            sink?(FlutterError(code: code, message: message, details: details))
        }
    }
}

/// Type of operation being performed.
enum OperationType: String {
    case download
    case load
}

/// Current status of the operation.
enum OperationStatus: String {
    case started
    case progress
    case completed
    case error
    case cancelled
}

/// Error codes for categorizing errors.
enum ErrorCode: String {
    case insufficientMemory = "INSUFFICIENT_MEMORY"
    case modelNotFound = "MODEL_NOT_FOUND"
    case networkError = "NETWORK_ERROR"
    case requestInterrupted = "REQUEST_INTERRUPTED"
    case modelLoadingFailure = "MODEL_LOADING_FAILURE"
    case internalError = "INTERNAL_ERROR"
    case unknown = "UNKNOWN"
}

/// Parses an error to determine its error code.
func parseErrorCode(from error: Error) -> ErrorCode {
    let errorString = "\(error)".lowercased()
    let localizedString = error.localizedDescription.lowercased()
    let combined = errorString + " " + localizedString

    // Check for memory-related errors
    if combined.contains("memory") ||
        combined.contains("oom") ||
        combined.contains("cannot allocate") ||
        combined.contains("allocation failed") ||
        combined.contains("out of memory")
    {
        return .insufficientMemory
    }

    // Check for request interrupted
    if combined.contains("interrupted") ||
        combined.contains("cancelled") ||
        combined.contains("canceled")
    {
        return .requestInterrupted
    }

    // Check for model not found
    if combined.contains("not found") ||
        combined.contains("does not exist") ||
        combined.contains("no such file")
    {
        return .modelNotFound
    }

    // Check for network errors
    if combined.contains("network") ||
        combined.contains("connection") ||
        combined.contains("timeout") ||
        combined.contains("unreachable")
    {
        return .networkError
    }

    // Check for internal errors
    if combined.contains("internalerror") ||
        combined.contains("internal error")
    {
        return .internalError
    }

    // Check for model loading failures
    if combined.contains("modelloadingfailure") ||
        combined.contains("failed to load")
    {
        return .modelLoadingFailure
    }

    return .unknown
}
